pkg/ddevapp/apptypes.go

```diff
@@ -0,0 +1,92 @@
+package ddevapp
+
+type settingsCreator func(*DdevApp) error
+type uploadDir func(*DdevApp) string
+
+// hookSuggestions should probably change its arg from string to app when
+// config refactor is done.
+type hookSuggestions func() []byte
+
+type siteSettingsPaths func(app *DdevApp)
+
+// AppTypeFuncs struct defines the functions that can be called (if populated)
+// for a given appType.
+type AppTypeFuncs struct {
+	settingsCreator
+	uploadDir
+	hookSuggestions
+	siteSettingsPaths
+}
+
+// AppTypeMatrix is a static map that defines the various functions to be called
+// for each apptype (CMS).
+// Example: AppTypeMatrix["drupal"]["7"] == { settingsCreator etc }
+var AppTypeMatrix = map[string]AppTypeFuncs{
+	"php": AppTypeFuncs{},
```

Comment:
For normal PHP applications, since we're not writing a settings file, where would a user look in order to find database connection details? Do we set environment variables that have this information? Or should we have some default place to write this data so that a user can just `require_once` a file and call it a day?

Reply:
The normal way to find out database settings is `ddev describe`.

Reply:
Right. My question is - are we okay explicitly saying that if you don't use one of our predefined app types, you're on your own for database configuration? Or can we help them out a little bit by making the credentials available somewhere in a machine readable way?

Reply:
I would say we're not committed to settings config at any point for any of these. Currently we only do it on db import, but I'm in favor of a new command that does it explicitly (if we have that functionality for the given apptype).  We have so many internal conflicts on settings.php on Drupal alone that we're quite stretched on how to do it. We haven't sorted out authoritatively what you do if settings.php exists. We create a settings.local.php, but without checking to see if settings.php includes it. And we don't know what to do if there's already a settings.local.php. It's kind of a rathole. Hard to solve for just Drupal.